crates/node/src/watchdog/mod.rs

```diff
@@ -0,0 +1,277 @@
+use eyre::Result;
+use hyper::server::conn::http1;
+use hyper::service::service_fn;
+use hyper::{Response, StatusCode};
+use hyper_util::rt::TokioIo;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tokio::net::TcpListener;
+use tokio::select;
+use tokio::sync::mpsc;
+use tokio::sync::Mutex;
+use tokio::time::timeout;
+use tokio::time::Duration;
+
+pub const GRACEFULL_SIGNAL_COUNTER_LIMIT: usize = 20;
+pub const SCHEDULER_HEATH_SIGNAL_TIMEOUT_MILLI: Duration = Duration::from_millis(1000);
+pub const NO_LOOP_DETECT_TIMEOUT_MILLI: Duration = Duration::from_millis(1000);
+
+#[derive(Clone, Debug, Copy, PartialEq, PartialOrd)]
+pub enum HealthCheckSignal {
+    SchedulerLoopOk,
+    SchedulerMempoolLen(usize),
+}
+
+#[derive(Clone, Debug, Copy, PartialEq, PartialOrd)]
+enum WatchDogState {
+    Alive,
+    Graceful,
+    Critical,
+}
+
+impl From<WatchDogState> for StatusCode {
+    fn from(state: WatchDogState) -> Self {
+        match state {
+            WatchDogState::Alive => StatusCode::OK,
+            WatchDogState::Graceful => StatusCode::NO_CONTENT,
+            WatchDogState::Critical => StatusCode::SERVICE_UNAVAILABLE,
+        }
+    }
+}
+
+pub async fn start_watchdog(bind_addr: SocketAddr) -> Result<mpsc::Sender<HealthCheckSignal>> {
+    let (scheduler_health_tx, scheduler_health_rx) = mpsc::channel::<HealthCheckSignal>(100);
+    let watchdog_state = Arc::new(Mutex::new(WatchDogState::Alive));
+    tokio::spawn({
+        let watchdog_state = watchdog_state.clone();
+        async move {
+            run_watchdog(
+                GRACEFULL_SIGNAL_COUNTER_LIMIT,
+                SCHEDULER_HEATH_SIGNAL_TIMEOUT_MILLI,
+                NO_LOOP_DETECT_TIMEOUT_MILLI,
+                watchdog_state,
+                scheduler_health_rx,
+            )
+            .await;
+        }
+    });
+
+    serve_files(bind_addr, watchdog_state).await?;
+    Ok(scheduler_health_tx)
+}
+
+async fn run_watchdog(
+    gracefull_signal_counter_limit: usize,
+    scheduler_heath_signal_timeout: Duration,
+    no_loop_detect_timeout: Duration,
+    watchdog_state: Arc<Mutex<WatchDogState>>,
+    mut scheduler_health_rx: mpsc::Receiver<HealthCheckSignal>,
+) {
+    let mut scheduler_state = (true, true, true);
```

Comment:
Maybe create a separate struct with named fields for this? I feel that it's a bit difficult to follow the code when these state fields are only referenced by the tuple index.